eth/gasestimator/gasestimator.go

```diff
@@ -0,0 +1,287 @@
+		return true, nil, err // Bail out
+	}
+	return result.Failed(), result, nil
+}
+
+// run assembles the EVM as defined by the consensus rules and runs the requested
+// call invocation.
+func run(ctx context.Context, call *core.Message, opts *Options) (*core.ExecutionResult, error) {
+	// Assemble the call and the call context
+	var (
+		evmContext = core.NewEVMBlockContext(opts.Header, opts.Chain, nil)
+		dirtyState = opts.State.Copy()
+	)
+	// Monitor the outer context and interrupt the EVM upon cancellation. To avoid
+	// a dangling goroutine until the outer estimation finishes, create an internal
+	// context for the lifetime of this method call.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Arbitrum: support NodeInterface.sol by swapping out the message if needed
+	var res *core.ExecutionResult
+	call, res, err := core.InterceptRPCMessage(call, ctx, dirtyState, opts.Header, opts.Backend, &evmContext)
+	if err != nil || res != nil {
+		return res, err
+	}
+
+	evm := opts.Backend.GetEVM(ctx, call, dirtyState, opts.Header, &vm.Config{NoBaseFee: true}, &evmContext)
+
+	go func() {
+		<-ctx.Done()
+		evm.Cancel()
+	}()
+	// Execute the call, returning a wrapped error or the result
+	gp := new(core.GasPool).AddGas(math.MaxUint64)
+	result, err := core.ApplyMessage(evm, call, gp)
+	if vmerr := dirtyState.Error(); vmerr != nil {
+		return nil, vmerr
+	}
+	if err != nil {
+		return result, fmt.Errorf("failed with %d gas: %w", call.GasLimit, err)
+	}
+
+	// Arbitrum: a tx can schedule another (see retryables)
+	scheduled := result.ScheduledTxes
+	for len(scheduled) > 0 {
+		// This will panic if the scheduled tx is signed, but we only schedule unsigned ones
+		msg, err := core.TransactionToMessage(scheduled[0], types.NewArbitrumSigner(nil), opts.Header.BaseFee)
+		if err != nil {
+			return nil, err
+		}
+		// The scheduling transaction will "use" all of the gas available to it,
+		// but it's really just passing it on to the scheduled tx, so we subtract it out here.
+		if result.UsedGas >= msg.GasLimit {
+			result.UsedGas -= msg.GasLimit
+		} else {
+			log.Warn("Scheduling tx used less gas than scheduled tx has available", "usedGas", result.UsedGas, "scheduledGas", msg.GasLimit)
+			result.UsedGas = 0
+		}
+		msg.TxRunMode = core.MessageGasEstimationMode
+		// make a new EVM for the scheduled Tx (an EVM must never be reused)
+		evm := opts.Backend.GetEVM(ctx, msg, dirtyState, opts.Header, &vm.Config{NoBaseFee: true}, &evmContext)
+		go func() {
+			<-ctx.Done()
+			evm.Cancel()
+		}()
+
+		scheduledTxResult, err := core.ApplyMessage(evm, msg, gp)
+		if err != nil {
+			return nil, err // Bail out
+		}
+		if err := dirtyState.Error(); err != nil {
+			return nil, err
+		}
+		if scheduledTxResult.Failed() {
+			return scheduledTxResult, nil
+		}
+		// Add back in any gas used by the scheduled transaction.
+		result.UsedGas += scheduledTxResult.UsedGas
+		scheduled = append(scheduled[1:], scheduledTxResult.ScheduledTxes...)
+	}
```

Comment:
This is almost the same as the code in `internal/ethapi/api.go`. Would it be easy to factor it out into a common function used by both to avoid issues of one being updated without the other one?

Reply:
I agree, this is a good idea!